Cloud Backend Test

ctgovea opened the pull request:
Hi, Carlos here (👋). The project is a 3-tier architecture on AppSync, Lambda and DynamoDB using the CDK v2. Environment variables to test will be provided separately. Cheers!

Commit:
chore: cdk init

Commit:
chore: add commitlint, husky

Commit:
feat: add graphql schema

Commit:
chore: rename bin entry file

Commit:
feat: add appsync dynamodb resolvers

Commit:
feat: auto populate timestamp properties

Commit:
feat: make id optional on createUser

Commit:
fix: comma separate attribute values

Commit:
chore: rename to id

Commit:
chore: rename to items

Commit:
chore: bump cdk version to 2.3.0

Commit:
feat: add location lambda

Commit:
chore: add dotenv package

Commit:
feat: add lambda to aws stack

ctgovea committed:
Merge pull request #1 from govea/ts-lambda

- Add a location lambda function in TypeScript
- Use Mapbox Places API to get the coordinates from a given address
- Add lambda to cloud stack

Commit:
chore: rename to queryAddress

Commit:
feat: add location query and type

Commit:
feat: add subscription on user updated

Commit:
feat: test cloud stack resources are created

Commit:
chore: refactor out schema.graphql

Commit:
chore: move lambdas folder under src

Commit:
chore: add prettier

Commit:
chore: pretty up lambda code

Commit:
feat: update documentation

Commit:
feat: update documentation

Commit:
Merge branch 'main' of https://github.com/govea/cloud-backend-test

Commit:
feat: add architecture diagram

Commit:
Revert "Merge branch 'main' of https://github.com/govea/cloud-backend-test"

This reverts commit c8b39836b995472f9a7e5aa256d1fed68c84d650, reversing
changes made to e4ca769bdda0286a76520442c011e84c81eb5917.

Commit:
chore: update README repo structure

Commit:
feat: add updateUser mutation

Commit:
feat: add updateUser mutation

Commit:
Merge branch 'main' of https://github.com/govea/cloud-backend-test

# Conflicts:
#	src/lambdas/user-update.ts

Commit:
feat: save updatedAt on update

Commit:
feat: update table removal policy for production

Commit:
feat: update documentation

Commit:
feat: update documentation

andretorres00123 commented on .gitignore:
```diff
@@ -0,0 +1,12 @@
+*.js
```
I guess you added `*.js` and `*.d.ts` because of how your typescript configuration is generating the JS files. Why not use 
 the [outDir](https://www.typescriptlang.org/tsconfig#outDir) compiler option to put all the compiled files in a single directory?

ctgovea replied:
True. This was just the default `gitignore`created when you bootstrap a new typescript project with the CDK. I think the reason they do it this way is because of the js files generated in the `bin` and `lib` folder. But, to your point, I like them being separate. Added dist folder and updated `.gitignore`

andretorres00123 commented on jest.config.js:
```diff
@@ -0,0 +1,8 @@
+  testMatch: ['**/*.test.ts'],
+  transform: {
+    '^.+\\.tsx?$': 'ts-jest'
+  }
```
How would you add tests coverage?

ctgovea replied:
Updated with 

`collectCoverageFrom: ['**/*.{ts}', '!**/*.d.ts', '!**/node_modules/**'],`

andretorres00123 commented on src/graphql/schema.graphql:
```diff
@@ -0,0 +1,51 @@
+type User {
```
How would add comments to document your schema?

ctgovea replied:
Comments can be added with triple and single quotes per the [spec](https://spec.graphql.org/October2021/#sec-Descriptions). Added a couple as example.

andretorres00123 commented on src/lambdas/get-location.ts:
```diff
@@ -0,0 +1,46 @@
+/*
+Return the coordinates of a given address using the MapBox Places API
+
+eg. provide an address to query = '1600 Pennsylvania Ave NW, Washington, DC'
```
I tried to test this address using Insomnia, but I am getting `null`. Do you know the reason?
![image](https://user-images.githubusercontent.com/46347630/148130436-bcd4d78a-bba5-4a67-acbb-3dc65e796ba1.png)

ctgovea replied:
Apparently it was a line break when constructing the mapbox url in the lambda. Should be fixed now.

andretorres00123 replied:
it is working now :slightly_smiling_face:

andretorres00123 commented on src/lambdas/get-location.ts:
```diff
@@ -0,0 +1,46 @@
+export const handler = async (event: any = {}): Promise<any> => {
+  let queryAddress = event.arguments.address
+  if (!queryAddress) {
+    return { statusCode: 400, body: `Error: You are missing the address argument` }
```
the `statusCode` property is not necessary as you are using AppSync and GraphQL. Errors should be mapped using the responseMappingTemplate

ctgovea replied:
You're right! This is not the ApiGateway :). Removed status codes and updated the response mapping templates for both lambdas.

andretorres00123 commented on src/lambdas/user-update.ts:
```diff
@@ -0,0 +1,55 @@
+
+const db = new AWS.DynamoDB.DocumentClient()
+
+export const handler = async (event: any = {}): Promise<any> => {
```
I think this lambda is not working when I test the `updateUser` mutation in your deployed API. Do you the reason?
![image](https://user-images.githubusercontent.com/46347630/148131997-1309f963-d0fd-4802-8674-9ef0eddc46c2.png)

ctgovea replied:
So the update seemed to be successful but the mutation was erroring out because it was not returning the id. The reason for this was the lambda was just returning the updated properties. Since id is never updated it was not being passed back to AppSync. Fixed lambda to return all user properties.

andretorres00123 commented on src/lambdas/get-location.ts:
```diff
@@ -0,0 +1,46 @@
+  * MAPBOX_API_TOKEN    = 'pk.token-goes-here'
+*/
+
+export const handler = async (event: any = {}): Promise<any> => {
```
How would be your strategy to unit tests this function?

ctgovea replied:
I would add test cases for the errors: missing address, mocking failure, not found; validating output payload.

andretorres00123 commented on src/lambdas/user-update.ts:
```diff
@@ -0,0 +1,55 @@
+import * as AWS from 'aws-sdk'
```
How would be your strategy to unit tests this file?

ctgovea replied:
I would refactor the dynamodb code into its own function, mock it, and test input (missing id, missing arguments, valid input) and output format. I would also validate the error responses contain a type and message prop so it could be handled by AppSync properly.

andretorres00123 commented on test/infrastructure/api-stack.test.ts:
```diff
@@ -0,0 +1,31 @@
+import * as cdk from 'aws-cdk-lib';
```
I tried to run this tests but I am getting the following error, any idea why?:
![image](https://user-images.githubusercontent.com/46347630/148132371-308a2fe0-37b9-4c74-933f-e8cd51512474.png)
![image](https://user-images.githubusercontent.com/46347630/148132417-50f8cbd4-1818-4e67-8d73-05757d5f4ab8.png)

andretorres00123 replied:
Are these tests running against the real resources?

ctgovea replied:
This looks like a path issue. I'll try to dig further.

andretorres00123 commented on src/lambdas/user-update.ts:
```diff
@@ -0,0 +1,55 @@
+  })
+
+  try {
+    await db.update(params).promise()
```
How would you encapsulate the calls to dynamodb, so in a way that if in the future we want to change the database it won't affect the handler logic?

ctgovea replied:
I would abstract away the dynamodb functionality into a single function. That way if we want to switch databases we can just change one line in the handler. And I would probably go a step further. This code to update an item in DynamoDB is pretty much generic. It can work for any entity/properties so I would probably encapsulate in its own function for reusability.

andretorres00123 reviewed (commented):
Hi @govea ! I have completed of review your PR. 
As feedback, I would have liked to see unit tests of the logic of the lambdas.

I will be looking forward to hearing your responses to my comments.
Good luck.

Commit:
chore: add comments to graphql schema

Commit:
fix: handle errors in AppSync

Commit:
feat: add test coverage

Commit:
chore: add dist folder

ctgovea commented:
Hi Andre, thanks for reviewing the PR. I have pushed some fixes including:

- Add dist folder
- Add test coverage
- Add sample comments to graphql schema
- Fix getLocation lambda
- Remove status codes from lambdas
- Add ResponseMappingTemplates on lambdas to handle errors
- Fix updateUser query by returning user properties in lambda

Let me know if you run into any other issue or have a question/comment. Thanks!

ctgovea committed:
Merge pull request #2 from govea/next

Add review fixes

Commit:
feat: add integration test on graphql api

Commit:
feat: add unit test on user update lambda

Commit:
feat: add more unit tests on update user lambda

ctgovea committed:
Merge pull request #3 from govea/next

Add unit and integration tests

ctgovea commented:
Few more updates: 

- Added example integration test to validate connectivity to the actual GraphQL API in AppSync.
- Added unit tests on the update user lambda.
